Moving storage logic from Persistence into Data classes

elrido opened the pull request:
Fixes #419 - As outlined there, we currently always store certain information on the filesystem in the configured data directory alongside the pastes, if Filesystem storage got configured (which is the default). The ServerSalt is always stored in the data directory, while the Traffic- and PurgeLimiter could be individually configured.

This branch
* Moves the storage logic from Persistence into Data classes
* Removed configurable `dir` for Traffic- and PurgeLimiter from the configuration - if still configured in the configuration file, they will get ignored
* For the existing Database storage, the old ServerSalt in the data directory gets migrated, then deleted. The *Limiters won't, as they have a very high turnover rate (they get replaced every 10 / 300 seconds, respectively, by default)
* Since the GoogleCloudStorage isn't yet in a release, we don't have to consider a migration of an existing in file ServerSalt.

Compared to the last release, our unit test coverage decreased slightly from 96.94% to 95.46%. I'll try to look into that next weekend, if possible, but don't think this should prevent a merge.

@rugk and @r4sas, now that all changes are in, please have a look and let us know of any things that need improvements.

elrido commented:
Please also see the description in  #812 on the improved GCS implementation.

elrido commented:
The code coverage is now back up to 96.35% and Scrutinizer and Codacy helped me find and improve some odd and dead code. 

The only thing that I'm not 100% sure about is this:
https://github.com/PrivateBin/PrivateBin/blob/9357f122b7c72bbf27ed0d018355b8d762decb97/lib/Data/GoogleCloudStorage.php#L293
@mvanholsteijn Scrutinzer had pointed out that this should return a string, so added a cast for that. But AFAIK we already store a string in there and that [method is supposed to return a string](https://github.com/googleapis/google-cloud-php-storage/blob/ca3817dae90d7d757381f3f63c9d7df58ca1416f/src/StorageObject.php#L549-L577), so I wonder, is the JSON decode even needed or is it a remain from before we used the namespace plus key?

mvanholsteijn commented:
If the setValue is always a string than the json encode and decode is not necessary.  I updated the code and made a PR https://github.com/PrivateBin/PrivateBin/pull/813

elrido commented:
I reviewed all the changes once more and updated some doc block and other comments, fixed some indentations and moved some constants around. Am now happy with the state.

@rugk and @r4sas, did any of you have the chance to review this yet?

r4sas commented:
I'll do review today bit later...

r4sas commented on lib/Configuration.php:
```diff
@@ -80,13 +80,11 @@ class Configuration
         'purge' => array(
             'limit'     => 300,
             'batchsize' => 10,
-            'dir'       => 'data',
```
Required for backward-compatability when importing value to database, etc.

elrido replied:
I decided not to implement a migration of purge and traffic limiter migration into database, only server salt and only if that is in the default location (data):
- purge limiter has a life time of 300s, so it immediately purges on the first paste or comment creation after the upgrade, but on the DB that is a very cheap operation (we query the paste table with a where condition on the expiration timestamp)
- traffic limiter has a life time of 10s, so within 10s after the upgrade is completely empty/to be purged anyway.

Apart from mentioning this change in the CHANGELOG.md, we should point this change out in the upgrade instructions, i.e. "If you use the (default) Filesystem storage and have changed the `dir` for either purge or paste limiter, move these files into the same `dir` as configured in the `model_options`. If you use the Database storage, make sure that the `config` table contains the `salt` row from the file `data/salt.php` - an automatic migration is attempted, but may fail if you used a non-standard `PATH`."

r4sas commented on lib/Configuration.php:
```diff
@@ -80,13 +80,11 @@ class Configuration
         'traffic' => array(
             'limit'      => 10,
             'header'     => null,
-            'dir'        => 'data',
```
Required for backward-compatability when importing value to database, etc.

r4sas reviewed (changes requested):
Current configuration can use different directories, so we must use them for import at least.

r4sas commented on tst/Bootstrap.php:
```diff
@@ -1,5 +1,11 @@
 <?php
 
+use Google\Cloud\Core\Exception\BadRequestException;
```
So all GCS stuff will be imported?

r4sas replied:
Didn't seen anything for janitor in issues for years, so it can be...

rugk commented on lib/Data/Filesystem.php:
```diff
@@ -223,6 +247,97 @@ public function existsComment($pasteid, $parentid, $commentid)
+                return self::_storeString(
+                    self::$_path . DIRECTORY_SEPARATOR . 'traffic_limiter.php',
+                    '<?php' . PHP_EOL . '$GLOBALS[\'traffic_limiter\'] = ' . var_export(self::$_last_cache, true) . ';'
+                );
```
What about a `default` blcok that throws an exception or so? – if the function is called with invalid parameters, IMHOP´ we can panick.

Or at least log an error…

This is to guard against improper usage of the functions given that PHP's type system is quite limited.

elrido replied:
It is really a code style question if we use exceptions or defined error values for error signaling (empty string, in this case - all storage operations are expected to return a non-empty value). All of the existing interfaces avoided exceptions, so I stuck to the same convention for the new methods.

Of course, I agree that errors do need to be dealt with, regardless if they are returned as `false`, an dedicated error value (empty string, -1), throw an exception or a result-type that either contains an error or the result (like go or rust do, for example).

So, do we handle the error condition at the moment? If the paste can't be read, we return false and this (you'll love the irony) raises an exception, caught in the controller:
https://github.com/PrivateBin/PrivateBin/blob/af54e70359ae5994abc9f505e33cea6fdccf639c/lib/Model/Paste.php#L36-L38
https://github.com/PrivateBin/PrivateBin/blob/af54e70359ae5994abc9f505e33cea6fdccf639c/lib/Controller.php#L317-L327

But the traffic limiter doesn't:
https://github.com/PrivateBin/PrivateBin/blob/af54e70359ae5994abc9f505e33cea6fdccf639c/lib/Persistence/TrafficLimiter.php#L175
I'm a bit on the fence, in this case - we should consider a logging system as this is a failure that should be logged, but doesn't prevent the user from creating the paste (we could read the file but not save it). The same goes for the other two persistence classes:
https://github.com/PrivateBin/PrivateBin/blob/af54e70359ae5994abc9f505e33cea6fdccf639c/lib/Persistence/ServerSalt.php#L68
https://github.com/PrivateBin/PrivateBin/blob/af54e70359ae5994abc9f505e33cea6fdccf639c/lib/Persistence/PurgeLimiter.php#L76

I think I saw the use of an error log mechanism in the GCS class that we could reuse for these cases, as a starting point?

elrido replied:
I added an error_log() call for these three cases, so errors storing these, after being read, can be found by the server operator in the php logs (they are not getting translated, as that would be the language of the visitor). In the case of the purge limiter, we skip the purge cycle, as in that case the limit would be likely outdated and potentially cause a purge to be triggered every time a paste gets created.

There are currently no unit tests covering that line - it could probably be exercised using the file storage and making the files read-only or such, after creating them.

rugk commented on lib/Data/Filesystem.php:
```diff
@@ -223,6 +247,97 @@ public function existsComment($pasteid, $parentid, $commentid)
+                        return self::$_last_cache[$key];
+                    }
+                }
+                break;
```
Again what about a default value…

IMHO just returning an empty string as a silent feedback is bad. Let's better panic if developers call functions improperly, so they fix it and the code does not silently fail.

rugk commented on lib/Data/GoogleCloudStorage.php:
```diff
@@ -217,6 +209,88 @@ public function existsComment($pasteid, $parentid, $commentid)
+            $o = $this->_bucket->object($key);
+            return $o->downloadAsString();
+        } catch (NotFoundException $e) {
+            return '';
```
Maybe log an error/warning at least?

Or is this error really expected?

ALso, if we ignore an exception, let's at least have a comment there that explains _why_ we do so.

mvanholsteijn replied:
AFAIK understood from the reference implementation (https://github.com/PrivateBin/PrivateBin/blob/1fd998f325d72ad0273c714ad39a441a41492730/lib/Data/Filesystem.php#L291), when the key is not found and empty string is returned.

The comment would be ` // no key was found` which seems a bit superfluous given the catch is for the NotFoundException.

rugk commented on tst/ModelTest.php:
```diff
@@ -102,6 +102,58 @@ public function testBasicWorkflow()
+        $statement->closeCursor();
+
+        $paste = $model->getPaste(Helper::getPasteId());
+        $paste->getDeleteToken();
```
Do I miss sth. or why do you get the delete token here, but do not use the return value whatsoever?

So either don't get it or test the deletion actually. (which would maybe be a new test, however)

elrido replied:
I probably added the line here to get coverage on the condition, that calls a get(), if the paste didn't get loaded yet. I could add a separate test to cover the same call separately, but the purpose here isn't to test that `getDeleteToken()`returns anything, just that calling it will cause the loading of the paste from storage.
https://github.com/PrivateBin/PrivateBin/blob/af54e70359ae5994abc9f505e33cea6fdccf639c/lib/Model/Paste.php#L177-L179

rugk replied:
…but if the remove the line the test suceeds anyway, does not it?

If so, then IMHO that's not a good test. It should fail if one does not do the thing it tries to assert…

elrido replied:
I don't see a way to get that section of code covered with any kind of assert and separating it would increase the test run cost (duplicating every part till that line), but be even less clear, wouldn't it? For reference, here it is in the public coverage report on the latest release:
https://privatebin.info/coverage/Model/Paste.php.html#178

elrido replied:
I did move it into an assertNotEmpty, mainly so there is a comment explaining what it is for.

rugk reviewed (changes requested):
Some small remarks. :slightly_smiling_face:
